### PR TITLE
CSHARP-2843: Disable TLS renegotiation when possible

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs [hunk 1]
```diff
@@ -54,8 +54,17 @@ public Stream CreateStream(EndPoint endPoint, CancellationToken cancellationToke
             {
                 var sslStream = CreateSslStream(stream);
                 var targetHost = GetTargetHost(endPoint);
+
+#if (NET6_0_OR_GREATER)
+                var options = GetAuthOptions(targetHost);
+                sslStream.AuthenticateAsClient(options);
+#elif (NETSTANDARD2_1_OR_GREATER)
+                var options = GetAuthOptions(targetHost);
+                sslStream.AuthenticateAsClientAsync(options, default).GetAwaiter().GetResult();
```
Comment (BorisDog) [inline src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs:63]: There is no sync version in net2_1. Also the [sslstream sync version fallbacks to async.](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs#L339)

**Review by papafe — APPROVED** [review]

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs (hunk 1 above)
Comment (JamesKovacs) [reply src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs:63]: The `cancellationToken` parameter should be passed into `AuthenticateAsClientAsync`, not a `default` one.

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs [hunk 2]
```diff
@@ -73,8 +82,15 @@ public async Task<Stream> CreateStreamAsync(EndPoint endPoint, CancellationToken
             {
                 var sslStream = CreateSslStream(stream);
                 var targetHost = GetTargetHost(endPoint);
+
+#if (NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER)
+                var options = GetAuthOptions(targetHost);
+                await sslStream.AuthenticateAsClientAsync(options, default).ConfigureAwait(false);
```
Comment (JamesKovacs) [inline src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs:88]: Same here. The `cancellationToken` parameter should be passed into `AuthenticateAsClientAsync`, not a `default` one.

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs [hunk 3]
```diff
@@ -106,6 +122,17 @@ private void DisposeStreamIgnoringExceptions(Stream stream)
             }
         }
 
+#if (NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER)
+        private SslClientAuthenticationOptions GetAuthOptions(string targetHost) => new()
```
Comment (JamesKovacs) [inline src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs:126]: We typically avoid abbreviations. `GetAuthenticationOptions` or `GetSslClientAuthenticationOptions`.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs]: Thanks! The older API didn't accept `cancellationToken`, so this was carried over.

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs (hunk 3 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs]: Done.

File: src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Core/Connections/SslStreamFactory.cs]: Thanks done.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**